### PR TITLE
Remove Blockstore manual compaction code

### PR DESCRIPTION
**Pull request by steviez**
#### Problem
The manual Blockstore compaction that was being initiated from LedgerCleanupService has been disabled for quite some time in favor of several optimizations.

#### Summary of Changes
Rip out the dead code (and eliminate one thread)

I was poking around LedgerCleanupService following an idea I discussed with Yueh-Hsuan, and then was reminded of the dead code in there. So, I decided to pick up some of the cleanup from https://github.com/solana-labs/solana/pull/27529. However, that PR modified compaction and purge routines. I thought it'd be more manageable to handle one item per PR, so here is just the compaction stuff. The compaction removal is more straightforward than the purge change as well, so this should be a less complex review.

### CODE REVIEW

File: core/tests/ledger_cleanup.rs [hunk 1]
```diff
@@ -331,23 +309,23 @@ mod tests {
...
     /// Check [`get_benchmark_config`] for the full list of arguments.
     ///
     /// Example command:
     /// BENCHMARK_SLOTS=1000000 BATCH_SIZE=1 SHREDS_PER_SLOT=25 NUM_WRITERS=8 \
-    /// PRE_GENERATE_DATA=false cargo test --release tests::test_ledger_cleanup_compaction \
+    /// PRE_GENERATE_DATA=false cargo test --release tests::test_ledger_cleanup \
     /// -- --exact --nocapture
     #[test]
-    fn test_ledger_cleanup_compaction() {
+    fn test_ledger_cleanup() {
```
Comment (steviez) [inline core/tests/ledger_cleanup.rs:323]: I went back and forth on whether to rip this test out (and thus this whole file):
- The bit about testing cleanup is now not really exercised here. Our cleaning methods (delete_files_in_range() for level OR fifo) both require a sizable amount of data to be inserted in order to see cleanup actually reclaim anything (on account of them operating on entire SST's). So, we'd have to make sure we pushed enough data to get several memtable flushes.
- We had used this as a bench previously for max shred insertion rate; however, I think the usefulness of that aspect may have gone away when merkle shreds were introduced and we could no longer hack the slot in quickly (thus making shred generation the bottleneck).

File: core/tests/ledger_cleanup.rs [hunk 2]
```diff
@@ -621,8 +593,8 @@ mod tests {
         );
         let u1 = storage_previous;
 
-        // Poll on some compaction happening
-        info!("Begin polling for compaction ...");
+        // Poll on some cleanup happening
+        info!("Begin polling for cleanup ...");
         let start_poll = Instant::now();
         while blockstore.storage_size().unwrap_or(0) >= u1 {
             if start_poll.elapsed().as_secs() > ROCKSDB_FLUSH_GRACE_PERIOD_SECS {
```
Comment (steviez) [inline core/tests/ledger_cleanup.rs:600]: Per my other comment, we may end up spinning this whole duration if we don't push enough data.

File: validator/src/main.rs [hunk 3]
```diff
@@ -3294,6 +3273,18 @@ fn get_deprecated_arguments() -> Vec<Arg<'static, 'static>> {
             .hidden(true)
             .takes_value(false)
             .help("Disable manual compaction of the ledger database (this is ignored)."),
+        Arg::with_name("rocksdb_compaction_interval")
```
Comment (steviez) [inline validator/src/main.rs:3276]: I moved the arguments to a retired function. This keeps them from cluttering the actual code without actually breaking anyone. 

It'd be nice to rip them completely; however, I'm fairly satisfied with them here such that they're no longer cluttering the actual code. Maybe we defer on ripping them out, and we can instead do a mass retiring of arguments at some point in the future if we really want to

File: ledger/src/blockstore/blockstore_purge.rs [hunk 4]
```diff
@@ -346,97 +339,6 @@ impl Blockstore {
                 .is_ok()
     }
 
-    pub fn compact_storage(&self, from_slot: Slot, to_slot: Slot) -> Result<bool> {
```
Comment (yhchiang-sol) [inline ledger/src/blockstore/blockstore_purge.rs:349]: Confirmed that we don't need RocksDB's `compact_range` (i.e. manual compaction) as the sole purpose of the manual compaction here is to clean up obsolete entries, and this is now done by delete_files_in_range in https://github.com/solana-labs/solana/pull/26651.

File: core/src/ledger_cleanup_service.rs [hunk 5]
```diff
@@ -287,39 +241,6 @@ impl LedgerCleanupService {
         Ok(())
     }
 
-    pub fn compact_ledger(
```
Comment (yhchiang-sol) [inline core/src/ledger_cleanup_service.rs:290]: 👍

File: core/tests/ledger_cleanup.rs (hunk 2 above)
Comment (yhchiang-sol) [reply core/tests/ledger_cleanup.rs:600]: This check is no longer valid as it expects to see the storage size shink w/o further writes, and this relies on manual compaction to exactly purge the entries older than `max_ledger_shreds`. 

I think a more proper check would be checking whether the number of shreds older than `last_slot - DEFAULT_PURGE_SLOT_INTERVAL` is smaller than `max_ledger_shreds`, but hope this isn't something tricky as it depends on the background activity.

File: core/tests/ledger_cleanup.rs (hunk 1 above)
Comment (yhchiang-sol) [reply core/tests/ledger_cleanup.rs:323]: Yep, agree that we can no longer abuse this test to do benchmarks due to the merkle shreds change.

We can decide the destiny of this file in a different PR.

File: core/tests/ledger_cleanup.rs [hunk 6]
```diff
@@ -645,74 +615,9 @@ mod tests {
...
-        }
-
         if config.cleanup_blockstore {
             drop(blockstore);
             Blockstore::destroy(&ledger_path).expect("Expected successful database destruction");
         }
     }
-
-    #[test]
-    fn test_compaction() {
```
Comment (yhchiang-sol) [inline core/tests/ledger_cleanup.rs:659]: 👍

File: validator/src/main.rs [hunk 7]
```diff
@@ -2308,10 +2294,6 @@ pub fn main() {
 
     let private_rpc = matches.is_present("private_rpc");
     let do_port_check = !matches.is_present("no_port_check");
-    let no_rocksdb_compaction = true;
```
Comment (yhchiang-sol) [inline validator/src/main.rs:2311]: Originally I was trying to find out whether we have arguments to change this, but I didn't see any related arguments.  This statement would be a confirmation that we have already hard-coded to have no manual compaction.

**Review by yhchiang-sol — COMMENTED** [review]
Thanks for working on this!  The removal of the manual compaction code looks good to me.

The only real comment is the ending check of the ledger_cleanup_test.

File: validator/src/main.rs (hunk 7 above)
Comment (steviez) [reply validator/src/main.rs:2311]: Correct, and for the sake of paper trail, here is the PR where this value was disconnected from validator arg and became constant:
https://github.com/solana-labs/solana/pull/16697/files#diff-59e2068dd4d49c21ab87f3a5d8792d2daca826ab6924b32413e551c0e3035f61L2019-R2019

File: core/tests/ledger_cleanup.rs (hunk 1 above)
Comment (steviez) [reply core/tests/ledger_cleanup.rs:323]: > We can decide the destiny of this file in a different PR.

Sounds good to me

File: core/tests/ledger_cleanup.rs (hunk 2 above)
Comment (steviez) [reply core/tests/ledger_cleanup.rs:600]: Marking this one as resolved - will respond in the other spot where you had this comment.

File: core/tests/ledger_cleanup.rs [hunk 8]
```diff
@@ -619,23 +589,22 @@ mod tests {
...
+        // have actually inserted that many slots or more.
+        let last_inserted_slot = benchmark_slots - 1;
+        let cleanup_service_may_have_run = last_inserted_slot > DEFAULT_PURGE_SLOT_INTERVAL;
+        if cleanup_service && cleanup_service_may_have_run {
+            let (exceeded_max_ledger_shreds, _, _) = LedgerCleanupService::find_slots_to_clean(
+                &blockstore,
+                last_inserted_slot - DEFAULT_PURGE_SLOT_INTERVAL,
+                max_ledger_shreds,
+            );
+            assert!(!exceeded_max_ledger_shreds);
```
Comment (steviez) [inline core/tests/ledger_cleanup.rs:606]: So here is the bit we were discussing, a few items:
- With the preset settings (50 slots), we will not enter this if statement
- The logic itself in the if statement is open to hit a race condition for the point you already mentioned of LedgerCleanupService running in the background
- I had to make `LedgerCleanupService::find_slots_to_clean` public in order to use it here to avoid duplicating the logic; also, `DEFAULT_PURGE_SLOT_INTERVAL` arguable shouldn't be public either

I know we said we could decide the fate of this test elsewhere, but I also don't want to spend more time / effort updating a test that we know is more or less obsolete. The test was designed to exercise compaction and could be used to bench insertion rate; compaction is gone and we know the bottleneck is in generation.

Thus, I'm inclined to do one of the following plans:
1) Rip this file out altogether, and add some unit tests in `ledger_cleanup_service.rs` in separate PR
2) Leave the test, but do NOT add this segment I have highlighted. This makes a function public that shouldn't be, and is essentially testing logic that could be tested on `find_slots_to_clean()` directly (hence my above point about unit tests in `LedgerCleanupService`. This would leave this test with no assertions

What do you think @yhchiang-sol ?

Comment (yhchiang-sol) [reply core/tests/ledger_cleanup.rs:606]: Great we share thoughts in common.  Here're my thoughts about this:

* Move forward with this PR.  Improvements/changes for the test go to a separate PR.
* This PR can probably remove the ending check of the test.  The test still has value as it can also warn us when the LedgerCleanupService did not clean up in time while the test is running (the soft and hard limit settings).

File: core/tests/ledger_cleanup.rs [hunk 9]
```diff
@@ -619,23 +589,22 @@ mod tests {
...
         }
-        info!(
-            "Done polling for compaction after {}s",
-            start_poll.elapsed().as_secs_f32()
-        );
-
-        let u2 = storage_previous;
```
Comment (yhchiang-sol) [inline core/tests/ledger_cleanup.rs:638]: Let's remove this check in this PR.  As the correct isn't something trivial to implement, let's put it in a separate PR.

**Review by yhchiang-sol — COMMENTED** [review]
Great we share thoughts in common.  Here're my thoughts about this:

* Move forward with this PR.  Improvements/changes for the test go to a separate PR.
* Remove the ending manual compaction check in the test in this PR.

File: core/tests/ledger_cleanup.rs (hunk 9 above)
Comment (steviez) [reply core/tests/ledger_cleanup.rs]: Done

**Review by yhchiang-sol — APPROVED** [review]
LGTM!

**Comment by steviez** [comment]
I think Ryo is currently preoccupied with some other more important stuff, but given that Ryo was the one who initially made the PR, pushing this in and will be happy to do address any potential issues in followup PRs
